src/paint/paint.c

```diff
@@ -59,17 +59,153 @@ bool sub_98196C(
 	uint16 z_offset,
 	uint32 rotation
 ) {
-	int flags = RCT2_CALLPROC_X(RCT2_ADDRESS(0x0098196C, uint32)[rotation],
-		x_offset | (bound_box_length_z << 8),
-		image_id,
-		y_offset,
-		z_offset,
-		bound_box_length_y,
-		bound_box_length_x,
-		rotation
-	);
-
-	return !(flags & (1 << 8));
+	assert((uint16) bound_box_length_x == (sint16) bound_box_length_x);
+	assert((uint16) bound_box_length_y == (sint16) bound_box_length_y);
+	assert((uint16) z_offset == (sint16) z_offset);
+
+	RCT2_GLOBAL(0xF1AD28, paint_struct*) = 0;
+	RCT2_GLOBAL(0xF1AD2C, uint32) = 0;
+
+	//Not a paint struct but something similar
+	paint_struct *ps = RCT2_GLOBAL(0xEE7888, paint_struct*);
+
+	if ((uint32) ps >= RCT2_GLOBAL(0xEE7880, uint32)) {
+		return false;
+	}
+
+	ps->image_id = image_id;
+
+	uint32 image_element = image_id & 0x7FFFF;
+	rct_g1_element *g1Element = gfx_get_g1_element(image_element);
+
+	rct_xyz16 coord_3d = {
+		.x = x_offset, // ax
+		.y = y_offset, // cx
+		.z = z_offset
+	};
+
+	rct_xyz16 boundBox = {
+		.x = bound_box_length_x, // di
+		.y = bound_box_length_y, // si
+		.z = bound_box_length_z,
+	};
+
+	switch (rotation) {
```

Comment:
I think these switches unrotate the variables. So rot 0 requires no rotation, rot 1 requires 3 rotations to get to 0, rot 2 requires 2, ... Can we add a comment stating that this rotates the variables so its relative to rotation 0.